op-verifier/contracts/lib/DisputeGameLookup.sol

```diff
@@ -0,0 +1,366 @@
+
+        // Get dispute game at index
+        Timestamp gameCreationTimeRaw;
+        (gameType, gameCreationTimeRaw, proxy) = disputeGameFactory.gameAtIndex(
+            index
+        );
+        outputRoot = proxy.rootClaim().raw();
+
+        // Unwrap gameCreationTime to uint64
+        gameCreationTime = gameCreationTimeRaw.raw();
+
+        // Wait for challenger to challenge the dispute game
+        if (block.timestamp - gameCreationTime < minAge) {
+            revert GameTooEarly(
+                index,
+                block.timestamp - gameCreationTime,
+                minAge
+            );
+        }
+
+        // Reject dispute game that has been expired
+        if (maxAge > 0 && block.timestamp - gameCreationTime > maxAge) {
+            revert GameExpired(
+                index,
+                block.timestamp - gameCreationTime,
+                maxAge
+            );
+        }
+
+        // Revert if the game is challenged
+        if (proxy.status() == GameStatus.CHALLENGER_WINS) {
+            revert GameChallenged(index);
+        }
+    }
+
+    /**
+     * @notice Retrieves the dispute game that is respected for L2 withdrawal.
+     * @param optimismPortal The Optimism portal output root contract.
+     * @param index The index of the dispute game.
+     * @param minAge The minimum age required to challenge the game.
+     * @param maxAge The maximum age allowed to challenge the game.
+     * @return outputRoot The root claim of the dispute game.
+     * @return gameType The type of the respected dispute game.
+     * @return gameCreationTime The creation time of the dispute game.
+     * @return proxy The dispute game proxy.
+     */
+    function getRespectedDisputeGame(
+        IOptimismPortalOutputRoot optimismPortal,
+        uint256 index,
+        uint256 minAge,
+        uint256 maxAge
+    )
+        internal
+        view
+        returns (
+            bytes32 outputRoot,
+            GameType gameType,
+            uint64 gameCreationTime,
+            IDisputeGame proxy
+        )
+    {
+        // Get gameType used for L2 withdrawal
+        GameType respectedGameType = optimismPortal.respectedGameType();
+
+        // Get dispute game output and game type
+        (outputRoot, gameType, gameCreationTime, proxy) = getDisputeGame(
+            optimismPortal,
+            index,
+            minAge,
+            maxAge
+        );
+
+        // Revert if gameType is not the one used for L2 withdrawal
+        if (gameType.raw() != respectedGameType.raw()) {
+            revert GameTypeMismatch(index, respectedGameType, gameType);
+        }
+    }
+
+    /**
+     * @notice Finds the search bound for binary search based on the maximum timestamp.
+     * @param disputeGameFactory The dispute game factory.
+     * @param maxTimestamp The maximum timestamp to search.
+     * @return lo The lower bound index.
+     * @return hi The upper bound index.
+     */
+    function _findSearchBound(
```

Comment:
These two binary searches are a clever approach when we don't know where to start!

I have an alternative suggestion that may prove simpler to implement and more efficient however: interpolation search. Start by fetching the timestamp of the oldest dispute game, and then take the ratio of `(now-maxTimestamp)/(now-oldestTimestamp)`. Use that to pick an intermediate element (instead of always picking the middle, as in binary search) and repeat as in binary search.

This should let us search the full space in one go, without spending a lot of time on values that are way too old as we would if we just did a binary search on the whole set of games.

It may help to set the midpoint to, say, 95% of what a naive evaluation would expect on the first iteration, to ensure the element we want usually falls within the smaller window.

Reply:
I have tried to implement this but it is very unstable compared to the two binary search approach

Like it doesn't work on some minimum time for an unknown reason in a random way

Reply:
![image](https://github.com/user-attachments/assets/e6181655-cb43-4b7a-8484-e4a942bd521c)

Old two binary search work on any high minimum time

![image](https://github.com/user-attachments/assets/3ffedb34-cce9-4b44-af15-8d7e5db6d3d4)

Interpolation search don't work on high minimum time

Reply:
Thanks for giving it a try! Did you try debugging it in Tenderly to see what's going wrong?